### ISSUE TITLE
Nintendo switch port of ppsspp (request).

### ISSUE THREAD

**Issue by iOS4all** [hrydgard/ppsspp#10934]
@hrydgard hi there.

Many people are wondering and excited about getting a Nintendo switch port of ppsspp (including me).
As you see Nintendo switch is already hacked specially on v3.0.0 we can run hombrews and emulators including retroarch on it. And also @plutoo is bringing JIT on LibNX (LibNX is the place to run hombrews). Also he and others developers are working to bring gpu drivers to libnx as well.

Finally @SciresM will bring CFW for Nintendo switch this summer.

It's very nice to get ppsspp for all platforms.

Thanks for your hard work.

**Comment by hrydgard** [comment]
Once there's a good homebrew environment for a publicly available CFW, and drivers for the GPU, a port should be fairly straightforward so I'm sure someone's gonna try it eventually. It's a bit too early still though.

**Comment by thedax** [comment]
This should be more feasible in the near future, as SDL2 has been ported, and a port of Mesa (and Nouveau) is being worked on for Horizon (the Switch's OS by Nintendo), so there will also be a standard OpenGL environment as well.

Keep an eye on https://github.com/devkitPro/libdrm_nouveau and https://github.com/devkitPro/mesa .

**Comment by iOS4all** [comment]
that’s fantastic news. Actually I saw someone on Twitter called Fincs he showed us app running on switch using OpenGL. I hope we can get ppsspp on horizon OS eventually.

**Comment by thedax** [comment]
Yeah, fincs is working on it. It's still very early though, so I think it's best to wait and see before jumping right in. As for the CFW situation, the Switch has been pretty hackable for the last 3 months or so via an exploit in the hardware's recovery mode, so full control is a non-issue on most Switch consoles (there are something like 15 or more million units vulnerable, so they're not hard to find currently). Debugging still leaves a bit to be desired, though.

**Comment by iOS4all** [comment]
Here’s a new news about OpenGL on switch see here:
![25bd579c-7239-475b-b297-f35ec5846938](https://user-images.githubusercontent.com/24762554/45130028-9307c600-b18f-11e8-86a6-eec25f732750.jpeg)

This is a great thing indeed.

To see his tweet see here
https://twitter.com/fincsdev/status/1037457485983096832?s=21

**Comment by thedax** [comment]
I'm keeping an eye on it. If the SDL2 port and this implementation of OpenGL work nicely together, that would mainly leave the recompiler (JIT). Horizon does provide JIT services if I remember correctly, so it should just be a matter of talking to that to get the ball rolling. Hopefully.

**Comment by iOS4all** [comment]
That’s something pretty nice to read.
About JIT I just want to correct you. Yes Plutoo brings JIT on LibNX officially while ago.

See here:
![732228a2-c542-4b0b-acf1-9046b2925d4a](https://user-images.githubusercontent.com/24762554/45131979-d9612300-b197-11e8-956b-9ee80368dcc4.jpeg)

Also here is an official GitHub link with JIT implemented.
https://github.com/switchbrew/libnx/blob/master/nx/include/switch/kernel/jit.h

**Comment by thedax** [comment]
What's there to correct? I said it does provide it if I remember correctly, and I did. :P

But yeah, I'll be waiting a bit longer, at least until the first public beta release.

**Comment by iOS4all** [comment]
Oh I’m sorry.
And all I can say good luck. Actually the Nintendo Switch really deserves to have ppsspp on it as a horizon OS especially after getting OpenGL. How nice.

**Comment by thedax** [comment]
It's fine, haha.

**Comment by hrydgard** [comment]
That means that all the big puzzle pieces are in place, at least. I'm personally not going to do Switch development until it gets a bit more stable and usable (is there already something like psplink? ) but there's nothing stopping anyone else from porting it. GL 4.3 should be alright.

**Comment by thedax** [comment]
You can get crash dumps via CFW and there are a couple WIP debuggers, but no gdb or stepping through code yet (to my knowledge).

**Comment by iOS4all** [comment]
LibNX is v1.4.0
With a huge improvement including OpenGL support.  finally.

For more info see here.
https://github.com/switchbrew/libnx/commit/23fa46a8ee6610c226ce7b555b204ca236e48623

I hope this could help about getting ppsspp on switch.

**Comment by thedax** [comment]
There are still several issues to work out before it'll build (I looked into it a bit earlier), but yeah, it's progressing quickly.

**Comment by thedax** [comment]
Alright, so I took a brief look at it last night, and here's what I came up with so far with regards to porting PPSSPP:

* [ ] There's no support for std::mutex or std::thread, so threading would have to be reworked or somehow implemented with Switch-specific stuff, maybe a wrapper around libnx that implements mutex and thread (is that even feasible? hm.)
* [ ] SDL2 port doesn't yet support OpenGL (SDL_GetError returns no GL driver support despite initializing via EGL port)
* [ ] Aarch64 JIT should work fine once hooked up to proper Switch-specific memory protection stuff
* [x] Build ppsspp-ffmpeg for Switch

**Comment by hrydgard** [comment]
I think std::mutex and std::thread could quite easily be implemented on top of Switch OS primitives, although that comes from looking at the 3DS OS, I haven't looked much at the Switch. That would be best done as part of libnx though.

GL support in the SDL2 port is probably just a matter of time, or a small patch...

**Comment by iOS4all** [comment]
@thedax there’s already a bounty for getting ARM64 JIT to work on n64 emu.
Also @natinusala said this can be works on other emulators as well.

Here’s a link to that bounty:
https://www.bountysource.com/issues/63766562-bounty-write-an-arm64-dynarec

I just want to to know about ARM64 JIT because you just mentioned it on your plans list.

**Comment by natinusala** [comment]
libnx has functions to create "JIT" buffers, e.g. buffers where the memory is executable. I looked at the PPSSPP code, and I think that we'll only need to hook it to `AllocateExecutableMemory`.

**Comment by thedax** [comment]
I saw the comments on that bountysource link, but for the sake of convenience: PPSSPP already has an ARM64 JIT like natinusala said on there. Getting the JIT to work is the least of our problems here. :P

**Comment by thedax** [comment]
Still waiting on SDL2 to be updated. Threading is another issue, but SDL2 is more important.

**Comment by natinusala** [comment]
SDL can easily be updated if you replace the GL includes by the glad includes and init the context properly. You can use mGBA or RetroArch as a reference for that, the switch-examples are borked.

**Comment by m4xw** [comment]
@thedax 
Make sure to recompile libdrm_nouveau since it doesnt contain the fixes via pacman yet, IIRC.
It should `just work` if set-up correctly.
There are still some bugs, like u can't re-load gland if u used eglTerminate for example.
I did the retroarch port via libnx, if you have questions feel free to ask me.

**Comment by iOS4all** [comment]
@thedax ???

**Comment by hrydgard** [comment]
@iOS4all Be patient, don't nag. If thedax wants to work on it, he will.

**Comment by iOS4all** [comment]
@hrydgard ok I’ll.

**Comment by dmiller423** [comment]
- figured I'd add some insight:

SDL2 has been updated and you can use GLES context it creates.

The wrapper is partially feasible, here is a very partial impl. for pthreads : https://github.com/m4xw/RetroArch/blob/fd155db9cca5696495440a35cc7fb9fc9b8831ff/libretro-common/rthreads/switch_pthread.h

unforunately it's nowhere near enough for the __gthread wrapper (libgcc std::thread wraps) and gthr-single.h just returns 0 or -1 to everything.

There are also various other issues,  mostly revolving around switch using a custom micro kernel that is not remotely posix compliant. IE: the only current method to map pages RW -> RX is show in the jit examples,  and thought it seems to be done in a single process now,  it gives back two address spaces:  one RW and one RX.  

I haven't looked into the details on Icache/block manager yet,  it might not be difficult to make the changes necessary to ensure the recompiler calls the 'second' (RX) address space.  

Likely you will run into many more,  the project is rather large and has multiple deps.

Here is a cmake file I used to test a bit,  since it's really a cross compile and I did not want to modify the base cmake file you need to configure like:   cmake .. -DCMAKE_TOOLCHAIN_FILE=./devkitA64.cmake

Below is the devkitA64.cmake

```
## devkitA64.cmake - devkitpro A64 cross-compile
#
set(CMAKE_SYSTEM_NAME Linux) # TODO Add CMake/Modules/Platform/switch 
set(CMAKE_SYSTEM_PROCESSOR aarch64)

set(CMAKE_SYSTEM_VERSION 1)   # this one not so much


set(DEVKITPRO $ENV{DEVKITPRO})
set(DEVKITA64 $ENV{DEVKITA64}) 
  
  
if ("" STREQUAL "${DEVKITPRO}")
  set(DEVKITA64 "/opt/devkitpro")
endif()

if ("" STREQUAL "${DEVKITA64}")
  set(DEVKITA64 ${DEVKITPRO}/devkitA64)
endif()


## specify the cross compiler
#
set(CMAKE_C_COMPILER   ${DEVKITA64}/bin/aarch64-none-elf-gcc)
set(CMAKE_CXX_COMPILER ${DEVKITA64}/bin/aarch64-none-elf-g++)


set(CMAKE_FIND_ROOT_PATH  ${DEVKITA64}) # where is the target environment

set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER) # search for programs in the build host directories
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)  # for libraries and headers in the target directories
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)




set(DKP_LIBNX        ${DEVKITPRO}/libnx)
set(DKP_LIBNX_INC    ${DKP_LIBNX}/include)

set(DKP_PORTLIBS     ${DEVKITPRO}/portlibs)
set(DKP_PORTLIBS_LIB ${DKP_PORTLIBS}/switch/lib)
set(DKP_PORTLIBS_INC ${DKP_PORTLIBS}/switch/include)

include_directories(${DKP_LIBNX_INC} ${DKP_LIBNX_INC}/switch/kernel ${DKP_PORTLIBS_INC})


#set(LIBZIP_INCLUDE_DIR)
#set(LIBZIP_LIBRARY 
#
set(ZLIB_INCLUDE_DIR    ${DKP_PORTLIBS_INC})      #${DEVKITPRO}/portlibs/switch/include)
set(PNG_PNG_INCLUDE_DIR ${DKP_PORTLIBS_INC})      #${DEVKITPRO}/portlibs/switch/include)
set(SDL2_INCLUDE_DIR    ${DKP_PORTLIBS_INC}/SDL2) #${DEVKITPRO}/portlibs/switch/include/SDL2)

set(ZLIB_LIBRARY        ${DKP_PORTLIBS_LIB}/libz.a)    #${DEVKITPRO}/portlibs/switch/lib/libz.a)
set(PNG_LIBRARY         ${DKP_PORTLIBS_LIB}/libpng.a)  #${DEVKITPRO}/portlibs/switch/lib/libpng.a)
set(SDL2_LIBRARY        ${DKP_PORTLIBS_LIB}/libSDL2.a) #${DEVKITPRO}/portlibs/switch/lib/libSDL2.a)



## gthr-single.h shouldn't be a valid options at least ?
#add_definitions(-D_GLIBCXX_HAS_GTHREADS -D__GTHREADS=1) ## *FIXME* build in wrapper, trick std::thread 
add_definitions(-D_GLIBCXX_USE_C99_MATH_TR1 -D_LDBL_EQ_DBL)
add_definitions(-DNSW -D__SWITCH__)


set(NSW ON)
set(ARM_NO_VULKAN ON)
set(USING_EGL ON)
set(USING_GLES2 ON)

```

**Comment by iOS4all** [comment]
Happy new year for all 🎊🎈

**Comment by DuIslingr** [comment]
Any updated progress on this?

**Comment by hrydgard** [comment]
I don't think anyone is working on it at this time. But as the Switch homebrew platform gets more and more complete, this will get easier and easier so I think there's a good chance it'll happen eventually.

**Comment by natinusala** [comment]
The latest requirement left is stf::thread :)

**Comment by iOS4all** [comment]
> The last requirement left is std::thread :)

How we can get this? Any idea please?

**Comment by natinusala** [comment]
Threads need to be implemented into devkitA64 (as a gthread implementation)

**Comment by dmiller423** [comment]
Threads aren't all that difficult,  ::Thread is a wrapper around gthreads,  which wrap pthread,  which in turn already has an implementation wrapping Thread on switch.

I also at least partially solved the issue with JIT read/write addresses, see: https://gist.github.com/dmiller423/1bac2ba090402417cfedda9122da25a9

However, there is no mmap() implementation unless this has changed since I last looked.  IIRC ppsspp used mmap quite extensively. There are direct mapping implementations of mmap()... putting the pieces together and making a full posix compliant mmap() or at least enough to satisfy ppsspp is a considerable challenge though.

**Comment by natinusala** [comment]
The pthread implementation we use in RetroArch isn't enough to implement everything in gthread I think

Isn't mmap only used to allocate executable memory ? That can be worked around with JIT buffers

**Comment by hrydgard** [comment]
We have a fallback solution, mmap mirroring is not required. No page guards are currently used in PPSSPP, unlike Dolphin for example.

**Comment by natinusala** [comment]
The Switch has a big issue though, it can't have RWX memory, only RW xor RX. 

We need to transition between RW and RX each time some code is written to a JIT buffer, a transition before writing it and one before executing it. If it slows down N64, it will definitely slow PSP down.

**Comment by iOS4all** [comment]
> The Switch has a big issue though, it can't have RWX memory, only RW xor RX.
> 
> We need to transition between RW and RX each time some code is written to a JIT buffer, a transition before writing it and one before executing it. If it slows down N64, it will definitely slow PSP down.

According to @m4xw he said if we patch the kernel we can get RWX. Of course without any slowdown. It’s likely to happen on atmosphere who knows.

**Comment by hrydgard** [comment]
Not an issue, we already implemented such a mode for iOS. Slowdown is fairly minor.

**Comment by m4xw** [comment]
@hrydgard You will need to be able to handle 2 bases and the overhead is compareable to a memcpy of the complete buffer.

**Comment by hrydgard** [comment]
Oh, right, different bases... yeah that complicates things. But can run without the JIT initially, some games will be playable with the IR interpreter I'm sure

**Comment by m4xw** [comment]
I mean, we have a workaround for that, but it's quite the overhead (we just got support for different bases in, still have to change it to use it).
So interpreter + underperforming dynarec it likely will be.

**Comment by dmiller423** [comment]
@m4xw can you provide metrics on the overhead?
And is this to cover the cost of changing the vm map back and forth to code memory?
( IE: the example minimal mprotect() I provided ) or another method?

**Comment by hrydgard** [comment]
@dmiller423  Search for PlatformIsWXExclusive()

**Comment by dmiller423** [comment]
Sorry I meant for the mmap() fallback you mentioned earlier

**Comment by hrydgard** [comment]
Set the flag MASKED_PSP_MEMORY in MemMap.h. That replaces the mirroring with masking. Then you can create a dummy MemArena implementation that allocates a giant buffer and have CreateView returns pointers offset into that.

**Comment by dmiller423** [comment]
Good bit of info,  I will check it out

**Comment by iOS4all** [comment]
> Good bit of info, I will check it out

Anything new about this please?

**Comment by hrydgard** [comment]
@iOS4all  If he has something to show, I'm sure he'll show it. Don't be impatient :)

**Comment by m4xw** [comment]
Patience isn't exactly the strength of this scene

**Comment by Emulatorer** [comment]
@iOS4all "Patience is a virtue"

**Comment by DuIslingr** [comment]
Atmosphere .8.4 added Jit kernel patches

Kernel patches for JIT support were added (Thanks, @m4xw!).
These loosen restrictions on caller process in svcControlCodeMemory.

**Comment by m4xw** [comment]
@hrydgard  Got it to compile after manually building a devkitA64 based toolchain with gthreads enabled newlib.
There are quite a few issues, glslang stuff references pthread, so that needs another toolchain update, my pthread wrapper isnt sufficient (or a rewrite using std::thread).
Lots of Issues like no ftello, fseeko, no ftruncate.
No gethostbyname, no pread, overall I had to change ~70 files (tho i think it can probably be cut in half with some better define placement).
Didn't touch JIT stuff yet.
Libretro makefile will need quite a few tweaks I think.
Not sure if I gonna release the toolchain, as I know devkitpro is working on their own solution.
Anyway thats the most common issues that appear, oh and also there are no dynamic libs, so need to rip dlsym stuff out either way (saw that in a few places).
Also glew (as is) will cause troubles.

**Comment by hrydgard** [comment]
Sounds like good progress. In addition to the missing threads, sounds like a bunch of issues to file in libnx then, as especially the file I/O functions would be nice to have there :) 

Yes we really should rip out glew for a number of reasons...

**Comment by m4xw** [comment]
I will take a look at what's required in newlib to enable them.

**Comment by m4xw** [comment]
Actually I just got told there might be a devkitA64 revision very soon, with std::thread and pthreads support.
So hopefully that should do the trick for glslang, meanwhile I can prepare the codebase, also offset based file api.

**Comment by azoreseuropa** [comment]
Sounds good. PPSSPP is a must for Nintendo Switch!

**Comment by hrydgard** [comment]
@m4xw  pthreads is indeed now available in libnx :)

**Comment by azoreseuropa** [comment]
> @m4xw pthreads is indeed now available in libnx :)

What does that mean ? :)

**Comment by iOS4all** [comment]
LibNX is now officially support pthreads and std::thread.

For more info see this tweet.
https://twitter.com/devkitpro/status/1111727942969761797?s=21

**Comment by m4xw** [comment]
@hrydgard I didn't find a way around missing mman yet.
I tried to do a custom implementation, but I couldn't map user data.
Only got the kernel data mapped at the wanted memory regions...
Alignment etc should be fine for the pages, so no idea why it would complain about invalid memory state in that case.
So I rewrote a whole bunch of memareasposix to work from heap and fix all the getBaseAddresses inline funcs etc to return them & a have the process manager service map the process into 32bit address space so it wouldn't malloc in 64bit addr space and truncate the pointers in the blockallocator
Anyway i consider that a giant hack and I guess that would cause a whole bunch of problems while emulating, maybe you have a better idea how this can be handled?

**Comment by iOS4all** [comment]
> > @m4xw pthreads is indeed now available in libnx :)
> 
> What does that mean ? :)

From my understanding these things are extremely important for ppsspp in order to run it.

**Comment by unknownbrackets** [comment]
Basically, if you imagine switch as a world, it previously didn't have gravity.  Now it has gravity.

PPSSPP is like a really awesome car, but it's kinda been built with the assumption that gravity will be working on the planet it's used on.

Now we're just working on the next problem, which is that the planet is -150 celcius and the car won't move because it's frozen.  So we're trying to figure out if we need to move a sun into place near the planet or if there's some other way to generate heat.

Once all of basic physics works like we thought it would, actually having PPSSPP work will probably not be too hard.

-[Unknown]

**Comment by m4xw** [comment]
Also another headsup, ofc their pthreads support had to be incomplete..

**Comment by m4xw** [comment]
https://twitter.com/m4xwdev/status/1113463278162862091
Got it to run ;)

**Comment by Aida-Enna** [comment]
Excellent work! inb4 "is it released yet" and "where can I download it". What remains to be worked on?

**Comment by m4xw** [comment]
Fix JIT, major cleanup, revise all changes, also hook up glsm properly (glew died in the battle).
Static linkage caused major issues for this

**Comment by azoreseuropa** [comment]
> Fix JIT, major cleanup, revise all changes, also hook up glsm properly (glew died in the battle).
> Static linkage caused major issues for this

Keep up the good work. Get your time and release it when you think it is ready. Amazing.

**Comment by hrydgard** [comment]
Awesome work m4xw! 

If you have managed to kill off glew, please submit that in a separate pull request before anything else. Been meaning to do that for ages, though I planned to simply grab and modify Dolphin's loader to fit us.

**Comment by m4xw** [comment]
Will do, first i need to do a **major** cleanup and some other updates for glsm.

**Comment by DuIslingr** [comment]
Any info regarding a standalone release for the switch?

**Comment by m4xw** [comment]
Standalone tease https://twitter.com/m4xwdev/status/1151131806055321603

**Comment by hrydgard** [comment]
Don't think we need this open anymore :)
